upf/plan_validator.py

```diff
@@ -0,0 +1,91 @@
+        self._simplifier = Simplifier(self._env)
+
+        self._substituter = Substituter(self._env)
+
+    def is_valid_plan(self, problem, plan) -> bool:
+        assignments = problem.initial_values()
+        for ai in plan.actions():
+            new_assignments: Dict[Expression, Expression] = {}
+            for ap, oe in zip(ai.action().parameters(), ai.parameters()):
+                assignments[ap] = oe
+            for p in ai.action().preconditions():
+                ps = self._subs_simplify(p, assignments)
+                if not (ps.is_bool_constant() and ps.bool_constant_value()):
+                    return False
+            for e in ai.action().effects():
+                cond = True
+                if e.is_conditional():
+                    ec = self._subs_simplify(e.condition(), assignments)
+                    if not (ec.is_bool_constant() and ec.bool_constant_value()):
+                        cond = False
+                if cond:
+                    ge = self._get_ground_fluent(e.fluent(), assignments)
+                    if e.is_assignment():
+                        new_assignments[ge] = self._subs_simplify(e.value(), assignments)
+                    elif e.is_increase():
+                        new_assignments[ge] = self._subs_simplify(self.manager.Plus(e.fluent(),
+                                                e.value()), assignments)
+                    elif e.is_decrease():
+                        new_assignments[ge] = self._subs_simplify(self.manager.Minus(e.fluent(),
+                                                e.value()), assignments)
+            assignments.update(new_assignments)
+            for ap in ai.action().parameters():
+                del assignments[ap]
+        for g in problem.goals():
+            gs = self._subs_simplify(g, assignments)
+            if not (gs.is_bool_constant() and gs.bool_constant_value()):
+                    return False
+        return True
+
+    def _get_ground_fluent(self, fluent:FNode, assignments: Dict[Expression, Expression]) -> FNode:
+        assert fluent.is_fluent_exp()
+        new_args = []
+        for p in fluent.args():
+            new_args.append(self._subs_simplify(p, assignments))
+        return self.manager.FluentExp(fluent.fluent(), tuple(new_args))
+
+    def _subs_simplify(self, expression: FNode, assignments: Dict[Expression, Expression]) -> FNode:
+        to_subst = True
+        new_exp = expression
+        while to_subst:
+        #This do-while loop is necessary because when we have a FluentExp with
+        #  some parameters, the first substitution substitutes the parameters with
+        #  the object: then every ground fluent is substituted with it's value.
+        #  It is a while loop because sometimes more than 2 substitutions can be
+        #  required.
+            old_exp = new_exp
+            new_exp = self._substituter.substitute(new_exp, assignments)
+            to_subst = not (old_exp == new_exp)
```

Comment:
```suggestion
        old_exp = None
        new_exp = expression
        while old_exp != new_exp:
        #This do-while loop is necessary because when we have a FluentExp with
        #  some parameters, the first substitution substitutes the parameters with
        #  the object: then every ground fluent is substituted with it's value.
        #  It is a while loop because sometimes more than 2 substitutions can be
        #  required.
            old_exp = new_exp
            new_exp = self._substituter.substitute(new_exp, assignments)
```